rmw_zenoh_cpp/src/impl/pubsub_impl.cpp

```diff
@@ -13,32 +13,46 @@ extern "C"
 
 std::mutex sub_callback_mutex;
 
-// Static message map
-std::unordered_map<std::string, std::vector<unsigned char> >
-  rmw_subscription_data_t::zn_messages_;
 
+/// STATIC SUBSCRIPTION DATA MEMBERS ===========================================
+size_t rmw_subscription_data_t::sub_id_counter = 0;
+
+// Map of Zenoh topic key expression to subscription data
+std::unordered_map<std::string, std::vector<rmw_subscription_data_t *> >
+  rmw_subscription_data_t::zn_topic_to_sub_data;
+
+
+/// ZENOH MESSAGE SUBSCRIPTION CALLBACK ========================================
 void rmw_subscription_data_t::zn_sub_callback(const zn_sample * sample) {
-  // Prevent race conditions...
   std::lock_guard<std::mutex> guard(sub_callback_mutex);
 
   // NOTE(CH3): We unfortunately have to do this copy construction since we shouldn't be using
   // char * as keys to the unordered_map
   std::string key(sample->key.val, sample->key.len);
 
-  // Vector to store the byte array (so we have a copyable type instead of a pointer)
+  // Vector to store the byte array (so we have a copyable container instead of a pointer)
   std::vector<unsigned char> byte_vec(sample->value.val, sample->value.val + sample->value.len);
 
-  // Fill the static message map with the latest received message
-  //
-  // NOTE(CH3): This means that the queue size for each topic is ONE for now!!
-  // So this might break if a topic is being spammed.
-  // TODO(CH3): Implement queuing logic
-  if (rmw_subscription_data_t::zn_messages_.find(key)
-      != rmw_subscription_data_t::zn_messages_.end()) {
-    // Log warning if message is clobbered
-    RCUTILS_LOG_WARN_NAMED(
-        "rmw_zenoh_cpp", "overwriting existing untaken zenoh message: %s", key.c_str());
+  // Get shared pointer to byte array vector
+  // NOTE(CH3): We use a shared pointer to avoid copies and to leverage on the smart pointer's
+  // reference counting
+  auto byte_vec_ptr = std::make_shared<std::vector<unsigned char> >(byte_vec);
```

Comment:
How should I go about doing this if I want to avoid a copy though? I can't make this variable a static variable because we have to deal with multiple topics with this one callback :O

Reply:
I consulted grey for this, apparently std::make_shared copies the data to the heap.

37c6dd30611e02e2875789affb3c2244ad198ea6 further removes this copy step by using `std::move()`. Then the data deallocation purely remains the responsibility of the shared_ptr (which will cause deallocation once all references are deleted)